lib/web/desktop.go

```diff
@@ -173,48 +174,175 @@ func proxyWebsocketConn(ws *websocket.Conn, con net.Conn) error {
+	return ps.wsOpen
+}
+
+// Close closes the websocket connection and sets wsOpen to false.
+// It should be deferred by all the goroutines that use playbackState,
+// in order to ensure that when one goroutine closes, all the others do too
+// (when used in conjunction with GetWsOpen).
+func (ps *playbackState) Close() {
+	ps.ws.Close()
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+	ps.wsOpen = false
+}
+
 func (h *Handler) desktopPlaybackHandle(
 	w http.ResponseWriter,
 	r *http.Request,
 	p httprouter.Params,
 	ctx *SessionContext,
 ) (interface{}, error) {
-	sID := p.ByName("session")
+	sID := p.ByName("sid")
 	if sID == "" {
-		return nil, trace.BadParameter("missing session in request URL")
+		return nil, trace.BadParameter("missing sid in request URL")
 	}
 
 	websocket.Handler(func(ws *websocket.Conn) {
-		defer ws.Close()
+		defer h.log.Debug("playback websocket closed")
+		rCtx := r.Context()
 		ws.PayloadType = websocket.BinaryFrame
+		ps := playbackState{
+			ws:      ws,
+			wsOpen:  true,
+			playing: true, // system always starts in a playing state
+		}
+		defer ps.Close()
 
-		var lastDelay int64
-		eventsC, errC := ctx.clt.StreamSessionEvents(r.Context(), session.ID(sID), 0)
-		for {
-			select {
-			case err := <-errC:
-				h.log.WithError(err).Errorf("streaming session %v", sID)
-				return
-			case evt := <-eventsC:
-				if evt == nil {
-					h.log.Debug("reached end of playback")
+		// Handle incoming playback actions.
+		go func() {
+			defer h.log.Debug("playback action-recieving goroutine returned")
+			defer ps.Close()
+
+			for {
+				if !ps.GetWsOpen() {
 					return
 				}
-				switch e := evt.(type) {
-				case *apievents.DesktopRecording:
-					if e.DelayMilliseconds > lastDelay {
-						time.Sleep(time.Duration(e.DelayMilliseconds-lastDelay) * time.Millisecond)
-						lastDelay = e.DelayMilliseconds
-					}
-					if _, err := ws.Write(e.Message); err != nil {
-						h.log.WithError(err).Error("failed to write TDP message over websocket")
+
+				action := playbackAction{}
+				// websocket.JSON.Receive will return even if there's no
+				// message waiting on the websocket, in which case action will
+				// just have it's default values (i.e. action.Action == "")
+				err := websocket.JSON.Receive(ws, &action)
+				if err != nil {
+					h.log.WithError(err).Error("error reading from websocket")
+					return
+				}
+				if action.Action != "" {
+					h.log.Debugf("recieved playback action: %+v", action)
+					if action.Action == actionToggle {
+						ps.TogglePlaying()
+					} else {
+						h.log.Errorf("received unknown action: %v", action.Action)
 						return
 					}
+				}
+			}
+		}()
+
+		// Stream session events back to browser.
+		go func() {
+			defer h.log.Debug("playback event-streaming goroutine returned")
+			defer ps.Close()
+
+			var lastDelay int64
+			eventsC, errC := ctx.clt.StreamSessionEvents(rCtx, session.ID(sID), 0)
+			for {
+				if !ps.GetWsOpen() {
+					return
+				}
+
+				if !ps.GetPlaying() {
```

Comment:
This will spin and waste CPU when we're paused. Any way we can wait on a channel or condvar when playback is paused instead?